### PR TITLE
Deterministic rotation/Hessian

### PR DESCRIPTION
**Pull request by f-dangel**
Background: I fixed the seed of `torch/scipy/numpy/random` and created a `quadratic_deep` test problem. Then I performed a forward pass. Then I did exactly the same a second time and expected the result of the forward pass to be the same. This was not true.

Reason: Creating two `quadratic_deep` problems results in different Hessians, as the random number generator used to generate them lives on the module level, and I could not 'reset' it.

This PR moves the random number generator to the functional level.

Demo:
- Before:  `h1` is different from `h2`
- After: `h1` is equal to `h2`

```python
import random

import numpy

import torch
from deepobs.pytorch.testproblems import quadratic_deep

# set seed
seed = 0
random.seed(seed)
numpy.random.seed(seed)
torch.manual_seed(seed)

# first problem
tp1 = quadratic_deep(5)
tp1.set_up()
h1 = tp1._hessian

# set seed
random.seed(seed)
numpy.random.seed(seed)
torch.manual_seed(seed)

# second problem
tp2 = quadratic_deep(5)
tp2.set_up()
h2 = tp2._hessian

# should be the same Hessian
print(torch.allclose(h1, h2))
```

### CODE REVIEW

**Comment by fsschneider** [comment]
Thanks!
We did not anticipate that someone would use two quadratic problems without restarting. But this PR fixes this.
Great!

**Comment by f-dangel** [comment]
Adding a mental note that the behavior is now different from the [tensorflow implementation](https://github.com/fsschneider/DeepOBS/blob/47fdaa3de1ea56a3143b8884c0f73f0ed85f199b/deepobs/tensorflow/testproblems/quadratic_deep.py#L10)
